src/dual.jl

```diff
@@ -765,6 +765,60 @@ function SpecialFunctions.logabsgamma(d::Dual{T,<:Real}) where {T}
     return (Dual{T}(y, SpecialFunctions.digamma(x) * partials(d)), s)
 end
 
+# Efficient left multiplication/division of #
+# Dual array by a constant matrix           #
+#-------------------------------------------#
+
+# creates the copy of x and applies fvalue!(values) to its values, and fpartial!(partial, ix) to its partials
+function _map_dual_components(fvalue!, fpartial!, x::AbstractArray{DT}) where DT <: Dual{<:Any, T} where T
+    N = npartials(DT)
+    @assert Base.elsize(x) == sizeof(T) * (N + 1) # check that accessing x as Vector{T} is possible
+    res = similar(x) # result
+    t = similar(x, T) # temporary Array{T} for fvalue!/fpartial! application
+    ystride = N + 1
+    # y allows res to be accessed as Vector{T}
+    y = Base.unsafe_wrap(Array, Base.unsafe_convert(Ptr{T}, res),
+                         length(res) * ystride, own=false)
+
+    # calculate res values
+    @inbounds for (j, v) in enumerate(x)
+        t[j] = value(v)
+    end
+    fvalue!(t)
+    k = 1
+    @inbounds for tt in t
+        y[k] = tt
+        k += ystride
+    end
+
+    # calculate each res partial
+    for i in 1:N
+        @inbounds for (j, v) in enumerate(x)
+            t[j] = partials(v, i)
+        end
+        fpartial!(t, i)
+        k = i + 1
+        @inbounds for tt in t
+            y[k] = tt
+            k += ystride
+        end
+    end
+
+    return res
+end
+
+Base.:\(m::Union{LowerTriangular{<:LinearAlgebra.BlasFloat},
+                 UpperTriangular{<:LinearAlgebra.BlasFloat},
+                 StridedMatrix{<:LinearAlgebra.BlasFloat}},
+        x::StridedVecOrMat{<:Dual}) =
+    _map_dual_components(Base.Fix1(ldiv!, m), (x, _) -> ldiv!(m, x), x)
+
+Base.:*(m::Union{LowerTriangular{<:LinearAlgebra.BlasFloat},
+                 UpperTriangular{<:LinearAlgebra.BlasFloat},
+                 StridedMatrix{<:LinearAlgebra.BlasFloat}},
+        x::StridedVecOrMat{<:Dual}) =
+    _map_dual_components(Base.Fix1(lmul!, m), (x, _) -> lmul!(m, x), x)
```

Comment:
I think it's going to be very hard to avoid ambiguities here. 
```
julia> rand(3,3) * Dual.(rand(3),0)
ERROR: MethodError: *(::Matrix{Float64}, ::Vector{Dual{Nothing, Float64, 1}}) is ambiguous.

Candidates:
  *(A::StridedMatrix{T}, x::StridedVector{S}) where {T<:Union{Float32, Float64, ComplexF32, ComplexF64}, S<:Real}
    @ LinearAlgebra ~/.julia/dev/julia/usr/share/julia/stdlib/v1.9/LinearAlgebra/src/matmul.jl:49
  *(m::Union{LowerTriangular{var"#s87", S} where {var"#s87"<:Union{Float32, Float64, ComplexF32, ComplexF64}, S<:AbstractMatrix{var"#s87"}}, UpperTriangular{var"#s86", S} where {var"#s86"<:Union{Float32, Float64, ComplexF32, ComplexF64}, S<:AbstractMatrix{var"#s86"}}, StridedMatrix{<:Union{Float32, Float64, ComplexF32, ComplexF64}}}, x::StridedVecOrMat{<:Dual})
    @ Main REPL[51]:4
```
Attaching the rule to `mul!(Matrix{<:Dual}, ...)` seems less likely to trigger them.

Testing with `Test.detect_ambiguities(ForwardDiff)` might be a good idea too.

Reply:
> Attaching the rule to `mul!(Matrix{<:Dual}, ...)` seems less likely to trigger them.

The problem is that promotion to `Dual` already happens at `\` (probably for `*` too).

Reply:
BTW, I think LinearAlgebra does something like this trick for some mixed real/complex cases (where the strides work out correctly). Maybe `mul!(ones(4,4).+im, rand(ComplexF64, 4,4), rand(4,4), true, false)` is one? Staying close to the signature used there is probably a way to avoid ambiguities.

Reply:
`\` might be easier than `*`, as it doesn't have so many methods.

For `*`, promotion to make `C` should happen correctly without this package doing anything, I think.

Reply:
> For `*`, promotion to make C should happen correctly without this package doing anything, I think.

This is [*(m::Triangular, x::Vector)](https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/triangular.jl#L1634)